File: plataform/MpOS Plataform Sharp/MpOS/Net/Rpc/RpcTcpServer.cs
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Reflection;
using Ufc.MpOS.Net.Rpc.Deploy.Model;
using Ufc.MpOS.Net.Rpc.Model;
using Ufc.MpOS.Net.Rpc.Util;
using Ufc.MpOS.Net.Util;
using Ufc.MpOS.Util.Exceptions;

namespace Ufc.MpOS.Net.Rpc
{
	public sealed class RpcTcpServer : TcpServer
	{
		private AssemblyLoader assembly;
		private readonly int bufferSize = 1460; //NOT change!!! this trash API didn't support packet...

		private readonly Type[] ReadMethodParamsTypes = { typeof(BinaryReader), typeof(string) };
		private readonly Type[] WriteMethodReturnTypes = { typeof(BinaryWriter), typeof(string), typeof(object) };

		public RpcTcpServer(string ip, RpcService service)
			: base(ip, service, typeof(RpcTcpServer))
		{
			startMessage = "Service: " + service.Name + "_" + service.VersionApp + "v, was started on port: " + service.Port;
			assembly = new AssemblyLoader(Service.Dependencies);
		}

		public override void ClientRequest(object connection)
		{
			TcpClient clientSocket = (TcpClient)connection;

			CallRemotable call = null;

			NetworkStream networkStream = clientSocket.GetStream();

			try
			{
				object objReturn = null;
				call = Receive(new BufferedStream(networkStream, bufferSize));

				if (call.Debug == null)
				{
					objReturn = InvokeMethod(call);
				}
				else
				{
					Stopwatch stopWatch = Stopwatch.StartNew();
					objReturn = InvokeMethod(call);
					stopWatch.Stop();

					call.Debug.ExecutionCpuTime = stopWatch.ElapsedMilliseconds;
				}

				Sent(new BufferedStream(networkStream, bufferSize), objReturn, call);
			}
			catch (Exception e)
			{
				if (call != null)
				{
					logger.Error("Method call: " + call.MethodName + ", from object: " + call.ObjLocal.GetType().Name + " failed!", e);
				}
				else
				{
					logger.Error("CallRemotable have some params which didn't loading correctly.", e);
				}

				SentError(new BufferedStream(networkStream, bufferSize), e);
			}
			finally
			{
				networkStream.ReadByte();

				Close(ref clientSocket, ref networkStream);
			}
		}

		private Object InvokeMethod(CallRemotable call)
		{
			int size = call.MethodParams.Length;
			Type[] typeParam = new Type[size];
			for (int i = 0; i < size; i++)
			{
				typeParam[i] = call.MethodParams[i].GetType();
			}

			MethodInfo method = call.ObjLocal.GetType().GetMethod(call.MethodName, typeParam);
			return method.Invoke(call.ObjLocal, call.MethodParams);
		}

		private CallRemotable Receive(BufferedStream stream)
		{
			//logger.Info(">> RpcTcpServer start Receive Data");
			CallRemotable call = new CallRemotable();

			byte[] headerData = new byte[5];
			stream.Read(headerData, 0, headerData.Length);
			int totalStreamSize = BitConverter.ToInt32(headerData, 0);

			if (headerData[4] == Code.CUSTOMSTREAM)
			{
				using (BinaryReader reader = new BinaryReader(ReadStream(stream, bufferSize, totalStreamSize)))
				{
					//logger.Info(">> RpcTcpServer Receive: build stream");
					string clsName = reader.ReadString();

					object objLocal = assembly.NewInstance(clsName);
					if (objLocal != null)
					{
						call.ObjLocal = objLocal;
						call.CustomSerialization = true;
						call.MethodName = reader.ReadString();

						call.MethodParams = (object[])objLocal.GetType().GetMethod("ReadMethodParams", ReadMethodParamsTypes).Invoke(objLocal, new object[] { reader, call.MethodName });
					}
					else
					{
						throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
					}
				}

			}
			else if (headerData[4] == Code.CUSTOMSTREAMDEBUG)
			{
				long totalDownloadTime = 0L;
				using (MemoryStream memoryStream = ReadStreamDebug(stream, bufferSize, totalStreamSize, ref totalDownloadTime))
				{
					call.Debug = new RpcProfile();
					using (BinaryReader reader = new BinaryReader(memoryStream))
					{
						string clsName = reader.ReadString();

						object objLocal = assembly.NewInstance(clsName);
						if (objLocal != null)
						{
							call.ObjLocal = objLocal;
							call.CustomSerialization = true;
							call.MethodName = reader.ReadString();
							call.MethodParams = (object[])objLocal.GetType().GetMethod("ReadMethodParams", ReadMethodParamsTypes).Invoke(objLocal, new object[] { reader, call.MethodName });

							call.Debug.DownloadSize = (int)memoryStream.Length;
							call.Debug.DonwloadTime = totalDownloadTime;
						}
						else
						{
							throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
						}
					}
				}
			}
			else if (headerData[4] == Code.BSONSTREAM)
			{
				using (BinaryReader reader = new BinaryReader(ReadStream(stream, 4096, totalStreamSize)))
				{
					string clsName = reader.ReadString();

					object objLocal = assembly.NewInstance(clsName);
					if (objLocal != null)
					{
						call.ObjLocal = objLocal;
						call.MethodName = reader.ReadString();
						call.MethodParams = GetParams(reader);
					}
					else
					{
						throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
					}
				}
			}
			else if (headerData[4] == Code.BSONSTREAMDEBUG)
			{
				long totalDownloadTime = 0L;
				using (MemoryStream memoryStream = ReadStreamDebug(stream, 4096, totalStreamSize, ref totalDownloadTime))
				{
					call.Debug = new RpcProfile();
					using (BinaryReader reader = new BinaryReader(memoryStream))
					{
						string clsName = reader.ReadString();

						object objLocal = assembly.NewInstance(clsName);
						if (objLocal != null)
						{
							call.ObjLocal = objLocal;
							call.MethodName = reader.ReadString();
							call.MethodParams = GetParams(reader);
							call.Debug.DownloadSize = (int)memoryStream.Length;
							call.Debug.DonwloadTime = totalDownloadTime;
						}
						else
						{
							throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
						}
					}
				}
			}

			//logger.Info(">> RpcTcpServer end Receive Data");
			return call;
		}

		private void Sent(BufferedStream stream, Object objReturn, CallRemotable call)
		{
			if (call.CustomSerialization)
			{
				if (call.Debug == null)
				{
					using (MemoryStream memoryStream = new MemoryStream(bufferSize))
					{
						BinaryWriter writer = new BinaryWriter(memoryStream);
						writer.Write(Code.CUSTOMSTREAM);

						call.ObjLocal.GetType().GetMethod("WriteMethodReturn", WriteMethodReturnTypes).Invoke(call.ObjLocal, new object[] { writer, call.MethodName, objReturn });
						writer.Flush();

						Sent(stream, memoryStream);
					}
				}
				else
				{
					using (MemoryStream memoryStream = new MemoryStream(bufferSize))
					{
						BinaryWriter writer = new BinaryWriter(memoryStream);
						writer.Write(Code.CUSTOMSTREAMDEBUG);
						writer.Write(call.Debug.DownloadSize);
						writer.Write(call.Debug.DonwloadTime);
						writer.Write(call.Debug.ExecutionCpuTime);
						call.ObjLocal.GetType().GetMethod("WriteMethodReturn", WriteMethodReturnTypes).Invoke(call.ObjLocal, new object[] { writer, call.MethodName, objReturn });
						writer.Flush();

						Sent(stream, memoryStream);
					}
				}
			}
			else
			{
				if (call.Debug == null)
				{
					using (MemoryStream memoryStream = new MemoryStream(bufferSize))
					{
						BinaryWriter writer = new BinaryWriter(memoryStream);
						writer.Write(Code.BSONSTREAM);

						byte[] bsonData = BsonFormatter.Serialize(objReturn);
						writer.Write(bsonData.Length);
						writer.Write(bsonData);
						writer.Flush();

						Sent(stream, memoryStream);
					}
				}
				else
				{
					using (MemoryStream memoryStream = new MemoryStream(bufferSize))
					{
						BinaryWriter writer = new BinaryWriter(memoryStream);
						writer.Write(Code.BSONSTREAMDEBUG);
						writer.Write(call.Debug.DownloadSize);
						writer.Write(call.Debug.DonwloadTime);
						writer.Write(call.Debug.ExecutionCpuTime);

						byte[] bsonData = BsonFormatter.Serialize(objReturn);
						writer.Write(bsonData.Length);
						writer.Write(bsonData);
						writer.Flush();

						Sent(stream, memoryStream);
					}
				}
			}
			//logger.Info(">> RpcTcpServer Sent Data");
		}

		private void Sent(BufferedStream stream, MemoryStream memoryStream)
		{
			byte[] data = memoryStream.ToArray();
			byte[] streamSize = BitConverter.GetBytes(data.Length);
			stream.Write(streamSize, 0, streamSize.Length);
			stream.Write(data, 0, data.Length);

			//logger.Info(">> RpcTcpServer Sent (data size): " + data.Length);
		}

		private void SentError(BufferedStream stream, Exception e)
		{
			using (MemoryStream memoryStream = new MemoryStream(512))
			{
				BinaryWriter writer = new BinaryWriter(memoryStream);

				writer.Write(Code.METHOD_THROW_ERROR);
				writer.Write(e.ToString());
				writer.Flush();

				Sent(stream, memoryStream);
			}
		}

		private object[] GetParams(BinaryReader streamReader)
		{
			int bsonLength = streamReader.ReadInt32();
			byte[] bsonData = new byte[bsonLength];
			streamReader.Read(bsonData, 0, bsonLength);
			return BsonFormatter.Deserialize(bsonData, assembly);
		}

		public new RpcService Service
		{
			get { return (RpcService)base.Service; }
		}
	}

	sealed class CallRemotable
	{
		public bool CustomSerialization { get; set; }
		public string MethodName { get; set; }
		public object[] MethodParams { get; set; }
		public object ObjLocal { get; set; }
		public RpcProfile Debug { get; set; }
	}
}

File: plataform/MpOS Plataform/src/br/ufc/mdcc/mpos/net/rpc/RpcTcpServer.java
/*******************************************************************************
 * Copyright (C) 2014 Philipp B. Costa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package br.ufc.mdcc.mpos.net.rpc;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;

import br.ufc.mdcc.mpos.net.TcpServerPool;
import br.ufc.mdcc.mpos.net.rpc.deploy.model.RpcService;
import br.ufc.mdcc.mpos.net.rpc.model.Code;
import br.ufc.mdcc.mpos.net.rpc.model.RpcProfile;
import br.ufc.mdcc.mpos.net.rpc.util.ClassLoaderException;
import br.ufc.mdcc.mpos.net.rpc.util.DebugBufferedInputStream;
import br.ufc.mdcc.mpos.net.rpc.util.DynamicClassLoaderEvent;
import br.ufc.mdcc.mpos.net.rpc.util.JarClassLoader;
import br.ufc.mdcc.mpos.net.rpc.util.RpcInputStream;
import br.ufc.mdcc.mpos.net.rpc.util.RpcSerializable;

/**
 * This server implementation was used for rpc (remote procedure call),
 * under TCP protocol.
 * 
 * @author Philipp B. Costa
 */
public final class RpcTcpServer extends TcpServerPool{
    private JarClassLoader jarLoader = null;
    private final int BUFFER_SIZE = 4096;// 4kb

    public RpcTcpServer(String cloudletIp, RpcService service) {
        super(cloudletIp, service, RpcTcpServer.class);
        startMessage = "Service: " + service.getName() + "_" + service.getVersionApp() + "v, was started on port: " + service.getPort();
        
        //only to get TcpServerPool class loader when start this thread 
        setDynamicClassLoaderEvent(new DynamicClassLoaderEvent() {       
            @Override
            public void setCurrentClassLoader(ClassLoader current) throws ClassLoaderException {
                jarLoader = new JarClassLoader(current, getService().getDependencies());
            }
        });
    }

    /**
     * Advice: All variables should be local to ensure thread-safe operations.
     */
    @Override
    public void clientRequest(Socket connection) throws IOException {
        CallRemotable call = null;
        Object objReturn = null;

        CloseableStream closeable = new CloseableStream();
        OutputStream output = connection.getOutputStream();
        InputStream input = connection.getInputStream();

        try {
            call = receive(input, closeable);
            if (call.debug == null) {
                objReturn = invokeMethod(call);
            } else {
                long initCpuTime = System.currentTimeMillis();
                objReturn = invokeMethod(call);
                call.debug.setExecutionCpuTime(System.currentTimeMillis() - initCpuTime);
            }

            send(output, closeable, objReturn, call);
        } catch (Exception e) {
            if (call != null) {
                logger.info("Method call: " + call.methodName + ", from object: " + call.objLocal.getClass().getName() + " failed!", e);
            } else {
                logger.info("CallRemotable have some params which didn't loading correctly. ", e);
            }
            sentError(output, closeable, e);
        } finally {
            waitClientClose(input);

            objReturn = null;
            call = null;

            close(closeable.dis);
            close(closeable.ois);
            close(closeable.dos);
            close(closeable.oos);

            closeable = null;
        }
    }

    private Object invokeMethod(CallRemotable call) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException, InstantiationException {
        Object serverObject = call.objLocal;

        for (Method method : serverObject.getClass().getMethods()) {
            if (method.getName().equals(call.methodName)) {

                Class<?>[] params = method.getParameterTypes();
                if (params.length == call.methodParams.length) {
                    try {
                        return method.invoke(serverObject, call.methodParams);
                    } catch (IllegalArgumentException e) {
                        // try call another overloaded method
                    }
                }
            }
        }
        return null;//not found...
    }

    @SuppressWarnings("resource")
    private CallRemotable receive(InputStream is, CloseableStream closeable) throws IOException, ClassNotFoundException, InstantiationException, IllegalAccessException {
        CallRemotable call = new CallRemotable();
        byte dataFlag[] = new byte[1];

        is.read(dataFlag);
        if (dataFlag[0] == Code.DATASTREAM) {
            DataInputStream dis = new DataInputStream(new BufferedInputStream(is, BUFFER_SIZE));

            String clsName = dis.readUTF();
            Object objLocal = jarLoader.newInstance(clsName);

            if (objLocal != null) {
                call.objLocal = objLocal;
                call.manualSerialization = true;
                call.methodName = dis.readUTF();
                call.methodParams = ((RpcSerializable) objLocal).readMethodParams(dis, call.methodName);
            } else {
                throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
            }
            closeable.dis = dis;
        } else if (dataFlag[0] == Code.OBJECTSTREAM) {
            ObjectInputStream ois = new RpcInputStream(jarLoader, new BufferedInputStream(is, BUFFER_SIZE));

            String clsName = ois.readUTF();
            Object objLocal = jarLoader.newInstance(clsName);

            if (objLocal != null) {
                call.objLocal = objLocal;
                call.methodName = ois.readUTF();
                call.methodParams = (Object[]) ois.readObject();
            } else {
                throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
            }
            closeable.ois = ois;
        } else if (dataFlag[0] == Code.DATASTREAMDEBUG) {
            call.debug = new RpcProfile();

            DebugBufferedInputStream bufInput = new DebugBufferedInputStream(is, BUFFER_SIZE);
            DataInputStream dis = new DataInputStream(bufInput);

            String clsName = dis.readUTF();
            Object objLocal = jarLoader.newInstance(clsName);

            if (objLocal != null) {
                call.objLocal = objLocal;
                call.methodName = dis.readUTF();

                call.manualSerialization = true;
                long totalDownloadTime = System.currentTimeMillis();
                call.methodParams = ((RpcSerializable) objLocal).readMethodParams(dis, call.methodName);
                call.debug.setDonwloadTime(System.currentTimeMillis() - totalDownloadTime);
                call.debug.setDownloadSize(bufInput.getTotalReadData());
            } else {
                throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
            }
            closeable.dis = dis;
        } else if (dataFlag[0] == Code.OBJECTSTREAMDEBUG) {
            call.debug = new RpcProfile();

            DebugBufferedInputStream bufInput = new DebugBufferedInputStream(is, BUFFER_SIZE);
            ObjectInputStream ois = new RpcInputStream(jarLoader, bufInput);

            String clsName = ois.readUTF();
            Object objLocal = jarLoader.newInstance(clsName);

            if (objLocal != null) {
                call.objLocal = objLocal;
                call.methodName = ois.readUTF();
                // avoid the initial lag and calcule param delay
                long totalDownloadTime = System.currentTimeMillis();
                call.methodParams = (Object[]) ois.readObject();
                call.debug.setDonwloadTime(System.currentTimeMillis() - totalDownloadTime);
                call.debug.setDownloadSize(bufInput.getTotalReadData());
            } else {
                throw new ClassNotFoundException("Update your dependence in side client, not found this class:" + clsName);
            }
            closeable.ois = ois;
        }
        return call;
    }

    private void sentError(OutputStream os, CloseableStream closeable, Exception e) throws IOException {
        byte dataFlag[] = new byte[1];

        dataFlag[0] = Code.DATASTREAM;
        os.write(dataFlag);

        DataOutputStream dos = new DataOutputStream(new BufferedOutputStream(os, BUFFER_SIZE));
        StringWriter sw = new StringWriter();
        e.printStackTrace(new PrintWriter(sw));

        dos.writeInt(Code.METHOD_THROW_ERROR);
        dos.writeUTF(sw.toString());
        dos.flush();

        closeable.dos = dos;
    }

    private void send(OutputStream os, CloseableStream closeable, Object objReturn, CallRemotable call) throws IOException {
        byte dataFlag[] = new byte[1];

        if (call.manualSerialization) {
            if (call.debug == null) {
                dataFlag[0] = Code.DATASTREAM;
                os.write(dataFlag);

                DataOutputStream dos = new DataOutputStream(new BufferedOutputStream(os, BUFFER_SIZE));
                dos.writeInt(Code.OK);
                ((RpcSerializable) call.objLocal).writeMethodReturn(dos, call.methodName, objReturn);

                dos.flush();
                closeable.dos = dos;
            } else {
                dataFlag[0] = Code.DATASTREAMDEBUG;
                os.write(dataFlag);

                DataOutputStream dos = new DataOutputStream(new BufferedOutputStream(os, BUFFER_SIZE));
                dos.writeInt(Code.OK);
                dos.writeInt(call.debug.getDownloadSize());//download on server perspective
                dos.writeLong(call.debug.getDonwloadTime());
                dos.writeLong(call.debug.getExecutionCpuTime());
                ((RpcSerializable) call.objLocal).writeMethodReturn(dos, call.methodName, objReturn);

                dos.flush();
                closeable.dos = dos;
            }
        } else {
            if (call.debug == null) {
                dataFlag[0] = Code.OBJECTSTREAM;
                os.write(dataFlag);

                ObjectOutputStream oos = new ObjectOutputStream(new BufferedOutputStream(os, BUFFER_SIZE));
                oos.writeInt(Code.OK);
                oos.writeObject(objReturn);

                oos.flush();
                closeable.oos = oos;
            } else {
                dataFlag[0] = Code.OBJECTSTREAMDEBUG;
                os.write(dataFlag);

                ObjectOutputStream oos = new ObjectOutputStream(new BufferedOutputStream(os, BUFFER_SIZE));
                oos.writeInt(Code.OK);
                oos.writeInt(call.debug.getDownloadSize());
                oos.writeLong(call.debug.getDonwloadTime());
                oos.writeLong(call.debug.getExecutionCpuTime());
                oos.writeObject(objReturn);

                oos.flush();
                closeable.oos = oos;
            }
        }
    }

    private void waitClientClose(InputStream input) {
        try {
            input.read(new byte[4]);
        } catch (IOException e) {
            logger.info("Some error during close socket!", e);
        }
    }

    @Override
    public RpcService getService() {
        return (RpcService) super.getService();
    }

    // wrapper class, for close stream correctly
    private final class CloseableStream {
        DataInputStream dis;
        DataOutputStream dos;
        ObjectInputStream ois;
        ObjectOutputStream oos;
    }

    private final class CallRemotable {
        boolean manualSerialization = false;
        String methodName;
        Object methodParams[];
        Object objLocal;
        RpcProfile debug;
    }
}